### DISCUSSION TITLE
Post-Processing Tools

### DISCUSSION THREAD

**Discussion by pgierz** [CMIP-Data-Request/CMIP7_DReq_Software#13] [General]
Dear CMIP7-DReq Team,

we (@esm-tools/cmor-dev-team: @mandresm, @siligam, and myself @pgierz) are currently working on building a Post Processing framework to correct metadata on simulation outputs (https://github.com/esm-tools/pymorize). It's still in a prototype state, but we would be interested in interfacing directly with any Python API that is made available to interact with the CMIP7 data requests. 

I had a brief look at what is available in the example workflow, which looks quite promising. I was curious if we could offer any assistance or collaboration work. Initially I think it would be good to provide some packaging around the scripts you currently provide so that they can be easily imported into other tools.

Cheers,
Paul

**Comment by JamesAnstey** [comment]
Dear Paul,

Thanks very much for taking a look at the developing CMIP7_DReq_Software and for letting us know about pymorize. I'd be interested to learn more about it and appreciate the offer to contribute/collaborate. 

Do you have suggestions for DReq API functions that would be needed to interface with pymorize? Would they interact with the `pymorize.controlled_vocabularies` or `pymorize.data_request` modules mentioned in the [pymorize API docs](https://pymorize.readthedocs.io/en/latest/API.html)? 

In general the DReq team would like to hear from modelling centres about how the data request can be integrated into their workflows, so any feedback on this topic is very welcome!

Best,
James

**Reply by pgierz** [reply]
Hi James,

yes that would be the general idea. From our side, I'd probably build a converter tool to interface whatever API the DReq functions provide to interface with our classes. I'd need to take a more detailed look at what is currently available, but we currently parse the JSON tables (at the moment, from CMIP6) into objects such that each requested variable corresponds to one of our [`DataRequestVariable`](https://pymorize.readthedocs.io/en/latest/api/pymorize.html#pymorize.data_request.DataRequestVariable) objects with attributes like full name, requested units, and so forth. We can then link our model output to each of these and start conversion processing. 

I'll have a discussion with my team and we can post some envisaged examples and use cases :-)

Cheers
Paul

**Reply by pgierz** [reply]
Hello again,

I have some time to come back to this now, after we finished integrating CMIP6 for our tool. At the moment, we have very closely modelled (read as: one to one) the classes we generate for interacting with the data request to the published CMIP6 Data Request JSON tables. I did some unnecessarily fancy abstract base class stuff so we can keep the interface the same once CMIP7 comes and I know which properties the new classes need to have.

So far, we have:

* `DataRequestVariable`: This is the basic building block object, and represents one particular variable found in one particular table. All the meta-data that might need to be in the final NetCDF file is somehow located on this objected as an attribute. In the end, I just want to be able to do something like this:

```python
import xarray as xr
ds = xr.open_dataset("our/model/output.nc")
drv = DataRequestVariable(...)
ds.attrs.update(**drv.attrs)
```

* `DataRequestTable`: these seem to no longer exist in CMIP7, unless I've overlooked something. There is still a way to figure out which CMIP6 Table name a variable comes from via the compound name, though. Here we had divided the object into two parts, a `Header` object, which had all the properties that apply to the whole table, and a list the `DataRequestVariable` objects described above.

* `DataRequest`: this is a sort-of "collection". For CMIP6, we just took all the JSON files in the provided folder and bundled them together. 

I went ahead and started a `pip install` version for this repository: #37. I guess what would be nice for downstream tools to have would be something like:

```python
>>> import cmip7_dreq  # Or whatever name makes most sense
>>> cmip7_dreq.get_variable("thetao", "Omon")  # Or Omon.thetao ? 
>>> cmip7_dreq.get_table("Omon")
[VariableA, VariableB, VariableC, ...]
>>> cmip7_dreq.get_opportunity(my_opportunity_name)
```

I suppose all the information is somewhere available, so it is just a matter of shaking it together in a nice way. 

I would be curious to see how other modelling centres have been handling this.

**Comment by JamesAnstey** [comment]
Hi Paul,

Thanks for your comments. It should be straightforward to make a function in the data request python API that returns the metadata for a single variable, like your DataRequestVariable object. There is some code doing essentially this to create the [all_var_info.json](https://github.com/CMIP-Data-Request/CMIP7_DReq_Software/blob/main/scripts/variable_info/all_var_info.json) file. 

Re. CMOR tables ("Omon", "Amon", etc) you're correct that these are not yet provided for CMIP7. Deciding how these will be organized is still a work in progress by the task team & WIP, and there's a desire to simplify if possible compared to CMIP6 (which had a lot of tables). However as you noted the Compound Name dating from CMIP6 is currently used as the unique name of a variable. These may not be the final names of variables but in any case a mapping will be provided so that one can go easily from the CMIP6-era name of a variable to its CMIP7-era name.

Thank you for starting a PR on a pip installer!

cheers
James
